button-merchant/src/main/java/com/usebutton/merchant/SSLManagerImpl.java

```diff
@@ -0,0 +1,163 @@
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ *
+ */
+
+package com.usebutton.merchant;
+
+import android.support.annotation.Nullable;
+import android.support.annotation.VisibleForTesting;
+
+import javax.net.ssl.KeyManagerFactory;
+import javax.net.ssl.SSLContext;
+import javax.net.ssl.TrustManagerFactory;
+
+import java.io.IOException;
+import java.security.GeneralSecurityException;
+import java.security.KeyManagementException;
+import java.security.KeyStore;
+import java.security.KeyStoreException;
+import java.security.NoSuchAlgorithmException;
+import java.security.cert.Certificate;
+import java.security.cert.CertificateException;
+import java.util.List;
+
+/**
+ * Network helper class that provides {@link KeyStore}s and {@link SSLContext} for use in secure
+ * socket communications.
+ */
+class SSLManagerImpl implements SSLManager {
+
+    private static SSLManager instance;
+
+    private KeyStore keyStore;
+    private SSLContext sslContext;
+    private Exception pendingException;
+    private List<Certificate> certificates;
+
+    private final char[] password;
+    private final CertificateProvider provider;
+
+    static SSLManager getInstance(CertificateProvider provider, @Nullable char[] password) {
+        if (instance == null) {
+            instance = new SSLManagerImpl(provider, password);
+        }
+
+        return instance;
+    }
+
+    @VisibleForTesting
+    SSLManagerImpl(CertificateProvider provider, @Nullable char[] password) {
+        this.password = password;
+        this.provider = provider;
+
+        init(provider);
+    }
+
+    @Override
+    public KeyStore getKeyStore()
+            throws CertificateException, NoSuchAlgorithmException, KeyStoreException, IOException,
+            KeyManagementException {
+        checkExceptions();
+        return keyStore;
+    }
+
+    @Override
+    public SSLContext getSecureContext()
+            throws CertificateException, NoSuchAlgorithmException, KeyStoreException, IOException,
+            KeyManagementException {
+        checkExceptions();
+        return sslContext;
+    }
+
+    @Override
+    public CertificateProvider getCertificateProvider() {
+        return provider;
+    }
+
+    KeyStore getKeyStore(CertificateProvider provider)
```

Comment:
Is this method only visible for testing? Does it need to be?

It seems like we should be able to avoid exposing the internal implementation by passing in a fake `CertificateProvider` that is compatible with the real implementation of this method, otherwise we're not really testing the code in this method.

Reply:
This was necessary because I was unable to create individual certificates for both the mock server and mock client. As such, I pass a previously created keystore which provides the certificates for both.

The functionality within this method is tested by the following tests:
`keyStore_shouldContainCertificates()`
`keyStore_shouldValidateCertificate()`

I will however mark this `@VisibleForTesting`